src/dbnode/storage/bootstrap/bootstrapper/fs/migrator/migrator.go

```diff
@@ -0,0 +1,261 @@
+	taskOptions    migration.TaskOptions
+}
+
+// Migrator is an object responsible for migrating data filesets based on version information in
+// the info files.
+type Migrator struct {
+	migrationTaskFn      MigrationTaskFn
+	infoFilesByNamespace fs.InfoFilesByNamespace
+	migrationOpts        migration.Options
+	fsOpts               fs.Options
+	instrumentOpts       instrument.Options
+	storageOpts          storage.Options
+	log                  *zap.Logger
+}
+
+// NewMigrator creates a new Migrator.
+func NewMigrator(opts Options) (Migrator, error) {
+	if err := opts.Validate(); err != nil {
+		return Migrator{}, err
+	}
+	return Migrator{
+		migrationTaskFn:      opts.MigrationTaskFn(),
+		infoFilesByNamespace: opts.InfoFilesByNamespace(),
+		migrationOpts:        opts.MigrationOptions(),
+		fsOpts:               opts.FilesystemOptions(),
+		instrumentOpts:       opts.InstrumentOptions(),
+		storageOpts:          opts.StorageOptions(),
+		log:                  opts.InstrumentOptions().Logger(),
+	}, nil
+}
+
+// migrationCandidate is the struct we generate when we find a fileset in need of
+// migration. It's provided to the workers to perform the actual migration.
+type migrationCandidate struct {
+	newTaskFn      migration.NewTaskFn
+	infoFileResult fs.ReadInfoFileResult
+	metadata       namespace.Metadata
+	shard          uint32
+}
+
+// mergeKey is the unique set of data that identifies an ReadInfoFileResult.
+type mergeKey struct {
+	metadata   namespace.Metadata
+	shard      uint32
+	blockStart int64
+}
+
+// completedMigration is the updated ReadInfoFileSet after a migration has been performed
+// plus the merge key, so that we can properly merge the updated result back into
+// infoFilesByNamespace map.
+type completedMigration struct {
+	key                   mergeKey
+	updatedInfoFileResult fs.ReadInfoFileResult
+}
+
+// Run runs the migrator
+func (m *Migrator) Run(ctx context.Context) error {
+	ctx, span, _ := ctx.StartSampledTraceSpan(tracepoint.BootstrapperFilesystemSourceMigrator)
+	defer span.Finish()
+
+	// Find candidates
+	candidates := m.findMigrationCandidates()
+	if len(candidates) == 0 {
+		m.log.Debug("no filesets to migrate. exiting.")
+		return nil
+	}
+
+	m.log.Info(fmt.Sprintf("found %d filesets to migrate. fileset migration start", len(candidates)))
+
+	nowFn := m.fsOpts.ClockOptions().NowFn()
+	begin := nowFn()
+
+	// Setup workers to perform migrations
+	var (
+		numWorkers = m.migrationOpts.Concurrency()
+		workers    = make([]*worker, 0, numWorkers)
+		outputCh   = make(chan completedMigration, len(candidates))
+	)
+
+	for i := 0; i < numWorkers; i++ {
+		// Give each worker their own persist manager so that we can write files concurrently.
+		pm, err := fs.NewPersistManager(m.fsOpts)
+		if err != nil {
+			return err
+		}
+		worker := &worker{
+			inputCh:        make(chan migrationCandidate, workerChannelSize),
+			outputCh:       outputCh,
+			persistManager: pm,
+			taskOptions: migration.NewTaskOptions().
```

Comment:
Ran `go test ./src/dbnode/storage/bootstrap/bootstrapper/fs/migrator/... -v -gcflags -m 2>&1 | grep migrator.go` and took a look at the output. `TaskOptions` were the only thing O(migrations) that were escaping to the heap, so made them a field on the worker so we can reuse. cc @robskillington

Reply:
I think we can do even better if the `TaskOptions` are created outside this loop (not a huge deal since it doesn't create anything large at the moment) 👍

Reply:
I think we have to create these in the loop so that each worker gets a unique set of `TaskOptions`. Otherwise, they'd clobber each other when they try to update some of the options when processing fileset.

Reply:
Oh nice! Will fix up